### ISSUE TITLE
`gno test -cover...`

### ISSUE THREAD

**Issue by moul** [gnolang/gno#1121]
Implement a basic code coverage support, and setup CI/CD to start monitoring the coverage status of the `examples/` folder.

**Comment by shashank-priyadarshi** [comment]
@michelleellen can I take this up?

**Comment by michelleellen** [comment]
@shashank-priyadarshi that sounds great, please check the bounty requirements and contributing guidelines referenced above. 
cc: @mvertes

**Comment by loren-osborn** [comment]
Hi @thehowl and @shashank-priyadarshi,

I’m interested in taking on #1121 if @shashank-priyadarshi is no longer interested in pursuing it. I saw his/your comment from March, but I haven’t seen any communication or updates since. If it’s no longer an active priority, I’d be happy to step in.

I’ve been a professional developer for many years—primarily in C++, PHP, MySQL, and build/automation tooling—with several years of Go experience. I’m new to the issue bounty ecosystem but very enthusiastic about contributing to meaningful projects. I’m especially passionate about test-driven development, and I view code coverage tools as essential for navigating and validating test completeness.

My plan is to follow the model used in Go, inserting lightweight counters into interpreted statements to track execution. Since this issue doesn’t involve gas tracking, I don’t believe a separate `GasMeter` is appropriate—in fact, I think preserving accurate gas accounting by avoiding interference is a key part of the design.

Are #1121, #2907, and #2908 still open for bounty participation? If so, could someone also point me to the most current documentation regarding bounty submission and payment terms?

Thanks for your time and for the work you’ve already put into this project—Gno is a fascinating platform, and I’d be grateful for the opportunity to contribute.

Best regards,
Loren Osborn

**Comment by shashank-priyadarshi** [comment]
Sure @loren-osborn , since offering to help, my focus shifted to other tasks, and I haven't been able to track this item. Thank you for taking it up!

**Comment by thehowl** [comment]
I don't think we have meaningful capacity to review any work on this ahead of the mainnet launch.

Let's consider this for after the mainnet launch.

**Comment by mudaly10-coder** [comment]
Auto-claim: taking this on (HAF).
Will submit a minimal PR quickly; happy to align scope/acceptance if needed.

**Comment by repairman29** [comment]
Hey there! This looks like something I could help with. I'd be happy to take a look and see what we can do.

*(Found via [Echeo](https://echeo.io))*

**Comment by jefft0** [comment]
Hi @notJoon . You opened PR https://github.com/gnolang/gno/pull/4241 to address this issue. Do you want to set yourself as an Assignee?

**Comment by notJoon** [comment]
@jefft0  let's check the status of the other related PRs first. if there's no progress, I'll go ahead and assign it to myself.

**Comment by thebrierfox** [comment]
Hi! I've submitted a PR for this bounty: #5156

Implementation highlights:
- Adds `-cover` flag to `gno test` with output matching Go's format: `coverage: XX.X%% of statements`
- Uses statement pointer identity from PackageNode.FileSet for O(1) hit tracking in the VM hot path
- 536 lines across 7 files, 10 unit tests, 1 integration test (coverage.txtar)
- All 46 existing txtar tests pass with zero regressions
- Tested on real packages (btree: 84.9%%, stack: 85.7%%, entropy: 83.9%%)

Happy to address any feedback quickly. Looking forward to review!

**Comment by energypantry** [comment]
I can take this now.\n\nDeliverable in <=24h (verifiable): open a PR with a failing coverage test and a first implementation patch for `gno test -cover` behavior.\nDelivery time (UTC): `2026-02-17T02:10:00Z`.\n\nPricing options:\nA) $420\nB) $780\nC) $1200\n\nAcceptance CTA: Reply A/B/C, or confirm budget and your preferred UTC start time. If approved, I start immediately.

**Comment by andygoodluck** [comment]
Hi @gnolang team,

I would like to claim this gno test -cover bounty ($4,000).

## Background

- 5+ years Go development with focus on tooling
- Experience with coverage tooling and VM instrumentation
- Contributed to testing infrastructure in production systems

## Implementation Plan

**Phase 1: GnoVM Instrumentation (Week 1)**
- Modify transpiler to inject coverage counters
- Match Go coverage semantics for statement detection

**Phase 2: gno test Integration (Week 1-2)**
- Implement -cover flag
- Match Go output format: coverage: 75.1% of statements
- Support -coverprofile for detailed reports

**Phase 3: CI/CD Integration (Week 2)**
- GitHub Actions for examples/ and gnovm/tests/
- Coverage badges and trend tracking

## Deliverables

- Coverage instrumentation in GnoVM
- -cover flag for gno test
- CI/CD integration
- Documentation and examples

Ready to start immediately!

GitHub: @andygoodluck

**Comment by thebrierfox** [comment]
@jefft0 Following up on this — it looks like the issue is still unassigned. As @notJoon mentioned, it's worth checking the status of the related PRs:

- #4241 — currently has merge conflicts, not mergeable
- #4570 — merge conflicts, no activity since Dec 2025
- #4973 — mergeable but no review engagement

My PR #5156 is mergeable, passes all 46 existing txtar tests, and covers the core requirement (`gno test -cover` with Go-matching output format). Happy to address any feedback or iterate quickly if there are design preferences.

Would you be available to take a look?

**Comment by huangpei2026psq** [comment]
GitHub username: huangpei2026psq
I'll take this. Thanks!

**Comment by MrLawrenceKwan** [comment]
Hi maintainers, I’d like to take bounty #1121 and start implementation now.\n\nPlan:\n1) instrument VM statement coverage counters (Go-like approach)\n2) add  output per package\n3) wire CI coverage checks for  and \n4) include tests + docs\n\nI’ll submit an incremental PR and keep updates in this thread.

**Comment by MrLawrenceKwan** [comment]
Correction to my previous comment: I am starting bounty #1121 now. Plan is VM coverage instrumentation + gno test cover output + CI coverage wiring + tests/docs, with incremental PR updates here.

**Comment by MrLawrenceKwan** [comment]
Progress update: opened PR #5160 implementing  with VM statement coverage tracking and package-level coverage output, plus test updates and CI wiring for gnotest coverage in examples/template and gnovm/tests.\n\nPR: https://github.com/gnolang/gno/pull/5160\n\nI could not run local Go test/gofmt in this environment because the  toolchain is unavailable, so CI will be the first full validation pass.

**Comment by MrLawrenceKwan** [comment]
Correction to previous note: PR #5160 implements gno test -cover with VM statement coverage tracking and package-level coverage output, plus tests and CI wiring for examples/template and gnovm/tests.\n\nPR link: https://github.com/gnolang/gno/pull/5160\n\nEnvironment note: I could not run local Go tests or gofmt here because the go toolchain is unavailable; CI will be the first full validation pass.

**Comment by thehowl** [comment]
Hi. We identified that over the past weeks, the bounty issues have attracted a steady stream of low-quality submissions using automated artificial intelligence tools. We are unable to currently process through these issues and don't consider these to be a valuable submission to our development community. As a consequence, issues and PR's related to the bounty program are no longer eligible. We'll update the contribution page on the gno.land website to reflect this. Contributors who contributed PR's that addressed bounties before September 2025 may still be eligible if there was a comment by a core team member specifically approving/validating the contribution.
